From cc0295ecd8befec2d94ad6719df3b52861e569c7 Mon Sep 17 00:00:00 2001
From: Chandrakala Chavva <Chandrakala.Chavva@cavium.com>
Date: Mon, 26 Oct 2020 19:22:09 -0700
Subject: [PATCH 671/767] driver: serdes_debugfs: Add inject optional parameter
 to prbs command

commit 49acfd1e2b1ac3c42e4d485fe24cec3c842f93b7 from
git@git.assembla.com:cavium/WindRiver.linux.git

If inject is passed to prbs command, then ATF will inject TX errors
on every RX errors.

Change-Id: Id8133270371a257c5ab3c0bee99b340e83d396aa
Signed-off-by: Chandrakala Chavva <Chandrakala.Chavva@cavium.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/38623
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Chandrakala Chavva <cchavva@marvell.com>
Integrated-by: Abhishek Paliwal <paliwal.abhishek@windriver.com>
---
 .../marvell/octeontx2-serdes/serdes_debugfs.c | 21 ++++++++++++++-----
 1 file changed, 16 insertions(+), 5 deletions(-)

diff --git a/drivers/soc/marvell/octeontx2-serdes/serdes_debugfs.c b/drivers/soc/marvell/octeontx2-serdes/serdes_debugfs.c
index 28725b49ba54..da32e3005b88 100644
--- a/drivers/soc/marvell/octeontx2-serdes/serdes_debugfs.c
+++ b/drivers/soc/marvell/octeontx2-serdes/serdes_debugfs.c
@@ -639,7 +639,7 @@ static const struct file_operations loop_serdes_dbg_settings_fops = {
 static int serdes_dbg_prbs_lane_parse(const char __user *buffer,
 				      size_t count, int *qlm,
 				      enum cgx_prbs_cmd *cmd, int *mode,
-				      int *qlm_lane)
+				      int *qlm_lane, int *inject)
 {
 	char *cmd_buf, *cmd_buf_tmp, *subtoken;
 	int ec;
@@ -682,6 +682,13 @@ static int serdes_dbg_prbs_lane_parse(const char __user *buffer,
 			subtoken = strsep(&cmd_buf, " ");
 			ec = subtoken ? kstrtoint(subtoken, 10, mode) :
 					-EINVAL;
+			if (ec == -EINVAL)
+				goto out;
+			subtoken = strsep(&cmd_buf, " ");
+			if (subtoken)
+				kstrtoint(subtoken, 10, inject);
+			else
+				*inject = 0;
 		} else if (!strcmp(subtoken, "stop")) {
 			*cmd = CGX_PRBS_STOP_CMD;
 		} else if (!strcmp(subtoken, "clear")) {
@@ -691,6 +698,7 @@ static int serdes_dbg_prbs_lane_parse(const char __user *buffer,
 		}
 	}
 
+out:
 	kfree(cmd_buf_tmp);
 	return ec;
 }
@@ -706,11 +714,13 @@ static ssize_t serdes_dbg_prbs_write_op(struct file *filp,
 	int qlm;
 	int ec;
 	int qlm_lane;
+	int inject;
 
 	ec = serdes_dbg_prbs_lane_parse(buffer, count, &prbs_cmd_data.qlm,
-					&cmd, &mode, &prbs_cmd_data.qlm_lane);
+					&cmd, &mode, &prbs_cmd_data.qlm_lane,
+					&inject);
 	if (ec < 0) {
-		pr_info("Usage: echo <qlm> <lane> [{start <mode>|stop}] > prbs\n");
+		pr_info("Usage: echo <qlm> <lane> [{start <mode> [inject]|stop|clear}] > prbs\n");
 		return ec;
 	}
 
@@ -719,8 +729,9 @@ static ssize_t serdes_dbg_prbs_write_op(struct file *filp,
 
 	switch (cmd) {
 	case CGX_PRBS_START_CMD:
-		arm_smccc_smc(OCTEONTX_SERDES_DBG_PRBS, cmd,
-			      qlm, mode, qlm_lane, 0, 0, 0, &res);
+		arm_smccc_smc(OCTEONTX_SERDES_DBG_PRBS, cmd, qlm,
+			      mode | (inject << 8),
+			      qlm_lane, 0, 0, 0, &res);
 
 		list_for_each_entry(status,
 				    &prbs_cmd_data.status_list.list,
-- 
2.31.1

